### ISSUE TITLE
Breaking change in URL encoding of OData key properties

### ISSUE THREAD

**Issue by lennart-m** [SAP/cloud-sdk-js#1398]
Hi, I found that the cloud SDK does handle URL encoding for key properties differently since 1.45.
I would like to know if this was intended as I did not find anything about this breaking change in the release notes.

The change is this: up to 1.44 the key properties were encoded when performing an OData GET.
Since 1.45 this does not happen anymore.

### Example:

I created an application for reading and writing BSP applications to ABAP repositories to use in our CI/CD pipelines, consuming the SAP standard service `/UI5/ABAP_REPOSITORY_SRV`. The entity `Repository` has a key consisting of only one property, `Name`.

In order to read information about a repository I need to perform this request:
```http
GET /sap/opu/odata/UI5/ABAP_REPOSITORY_SRV/Repositories(Name='%2FUI2%2FUSHELL')?$format=json
```
When using version 1.44 of the cloud SDK exactly this URL is generated.

But since 1.45 the request looks like this:
```http
GET /sap/opu/odata/UI5/ABAP_REPOSITORY_SRV/Repositories(Name='/UI2/USHELL')?$format=json
```
This is a valid URL consisting of the segments 
- `Repositories(Name='`
- `UI2`
- `USHELL')?$format=json`

where there should only be one segment
- `Repositories(Name='%2FUI2%2FUSHELL')?$format=json`

So the system responds with this message:
_Invalid URI segment 'Repositories(Name=''_

Is this behavior intended? In this case the calling application would have to do the encoding itself; but in this case the cloud SDK must not do additional encoding in order to prevent duplicate encoding of strings.

Thanks a lot,

Lennart

**Comment by jjtang1985** [comment]
Hi @lennart-m ,

Thanks for approaching us.
I assume you use the code snippets like 
```ts
YourEntity
.requestBuilder()
.getByKey(`/UI2/USHELL`)
```
and I can reproduce the issue, where the `/` is not encoded.

I'll create a backlog for it.

May I ask your timeline and do you have a workaround for it?

Best regards,
Junjie

**Comment by lennart-m** [comment]
Hi @jjtang1985,

Thanks for your reply. You're right this is the URL I use. I haven't tested using `update` or `delete` but I guess URL generation will show the same behavior for those as well.

We are using the cloud SDK for internal purposes only (deployment in CI/CD) so there is no urgency. The workaround is not to upgrade the dependencies to 1.45 but stick to 1.44 or lower and this works fine.

Thanks a lot,

Lennart

**Comment by FrankEssenberger** [comment]
Hello,

I just saw the issue and I broke this in this PR: https://github.com/SAP/cloud-sdk-js/pull/1283/files#diff-904c154f1add39f6544ba9447d4117f630274ae39e123ceae54fc517e4dde352

The aim was to encode the URL only once and I thought that the filter was the only user provided part where encoding was needed. I forgot about the getByKey values. So the encoding should be also added there.

**Comment by jjtang1985** [comment]
Hi @lennart-m ,

we fixed the bug and you can try the `canary` version for now.
Please provide some feedbacks if possible.
We'll release a new stable version soon. (hopefully this week)

Best regards,
Junjie

**Comment by marikaner** [comment]
Hey @lennart-m,

this issue should be fixed in version 1.48.0. I will go ahead and close this issue for now, but let us know if you encounter further problems. Thank you for reporting this!

**Comment by lennart-m** [comment]
Hi, I tested this in the last week and it seems like nothing has changed since 1.45. The OData key is still not escaped, resulting in _Invalid URI segment_ errors.
I tested with 1.48 and 1.53, using both a normal and a batch request.

Would you mind checking if the fix is really included in the current version?

Until then I will continue to use 1.44 since it's the last working version.

**Comment by marikaner** [comment]
Hey @lennart-m,

that is unfortunate. We discussed this issue today and will work on this soon. However, we will fix this in version 2.x of the SDK. We will let you know once this is resolved.

**Comment by deekshas8** [comment]
Hi @lennart-m,

We have fixed the bug in our latest released v2.1.0. Let us know if this resolves your issue.
I'll close the ticket for now. If you still experience problems, please reopen.

Note: We released our new major 2.0 version earlier this month and will no longer be fixing bugs in version 1.x. Please upgrade to v2 of the SDK. You can find the [upgrade guide](https://sap.github.io/cloud-sdk/docs/js/guides/upgrade-to-version-2) for the steps. If you face issues during the upgrade, feel free to open an issue.
